mqtt/mqtt-bridge/src/token_source.rs

```diff
@@ -29,50 +31,88 @@ impl SasTokenSource {
                 let expiry = expiry.timestamp().to_string();
                 let audience = format!(
                     "{}/devices/{}/modules/{}",
                     provider_settings.iothub_hostname(),
-                    provider_settings.device_id(),
-                    provider_settings.module_id()
+                    percent_encode(provider_settings.device_id().as_bytes(), IOTHUB_ENCODE_SET)
+                        .to_string(),
+                    percent_encode(provider_settings.module_id().as_bytes(), IOTHUB_ENCODE_SET)
+                        .to_string()
                 );
                 let resource_uri =
                     percent_encode(audience.to_lowercase().as_bytes(), IOTHUB_ENCODE_SET)
                         .to_string();
                 let sig_data = format!("{}\n{}", &resource_uri, expiry);
 
-                let client = edgelet_client::workload(provider_settings.workload_uri())
-                    .map_err(TokenSourceError::CreateClient)?;
+                let client =
+                    edgelet_client::workload(provider_settings.workload_uri()).map_err(|e| {
+                        Error::new(
+                            ErrorKind::Other,
+                            format!("could not create workload client: {}", e),
+                        )
+                    })?;
                 let signature = client
                     .sign(
                         provider_settings.module_id(),
                         provider_settings.generation_id(),
                         &sig_data,
                     )
                     .await
-                    .map_err(TokenSourceError::Sign)?;
+                    .map_err(|e| {
+                        Error::new(ErrorKind::Other, format!("could not get signature: {}", e))
+                    })?;
                 let signature = signature.digest();
-                UrlSerializer::new(format!("sr={}", resource_uri))
+                let token = UrlSerializer::new(format!("sr={}", resource_uri))
                     .append_pair("sig", &signature)
                     .append_pair("se", &expiry)
-                    .finish()
+                    .finish();
+
+                Some(format!("SharedAccessSignature {}", token))
             }
-            Credentials::PlainText(creds) => creds.password().into(),
-            Credentials::Anonymous(_) => "".into(),
+            Credentials::PlainText(creds) => Some(creds.password().into()),
+            Credentials::Anonymous(_) => None,
         };
 
         Ok(token)
     }
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum TokenSourceError {
-    #[error("failed to save to store.")]
-    Sign(#[from] edgelet_client::WorkloadError),
+#[derive(Clone)]
+pub struct TrustBundleSource {
+    creds: Credentials,
+}
+
+impl TrustBundleSource {
+    pub fn new(creds: Credentials) -> Self {
+        Self { creds }
+    }
+
+    pub async fn get_trust_bundle(&self) -> Result<Option<String>, Error> {
```

Comment:
Similar question as above. Why ```Option<String>```? In what cases would the trusted CA be None? Isn't this an error?

Reply:
if there is no workload api it doesn't get any certs, and it should be fine if upstream has a cert signed by a known CA